File: myflames/teach/join_family/semijoin_firstmatch.py
"""Lesson: Semijoin FirstMatch — the early-out on the first matching row.

The flagship of the semijoin cluster. A ``WHERE col IN (subquery)`` or
``EXISTS (subquery)`` is a *does-it-exist* question, not a *how-many*
question. The optimizer rewrites it as a semijoin and, with the
FirstMatch strategy, stops scanning the inner side the instant one
matching row is found for the current outer row — the inner scan
short-circuits.

Two acts drive the punchline (agreed with the teaching skill owner):

* **Act 1 — the naive mental model**: a plain inner join scans *every*
  matching order for each customer. If Alice has three orders over
  $1000, all three are read. That is wasted work for ``IN``.
* **Act 2 — FirstMatch early-out**: the same query, executed as a
  semijoin. The moment the *first* qualifying order is found for a
  customer, MySQL emits that customer and skips straight to the next
  customer. The later orders are never read — greyed out in the
  animation. The counters diverge: rows-read-naive climbs, rows-read
  FirstMatch flattens.

Structural template: ``nested_loop.py`` (the modern flagship pattern —
``_html.load_lesson_js`` sibling ``.js``, f-string HTML, ``anim.path``
arc pills, phase-nav, tweened arrivals, interactive complexity chart).

FirstMatch source note (route to mysql-correctness-reviewer):
the strategy lives in ``sql/sql_executor.cc`` /
``sql/iterators/composite_iterators`` as the ``FirstMatch`` semijoin
strategy; enabled by ``optimizer_switch='firstmatch=on'`` (default on)
in MySQL 8.4 and MariaDB 11.4. The claim animated here is behavioural
("scanning of the inner table for the current outer row stops after the
first match"), which should be checked against source before shipping
any hard number.
"""
from .. import _html


_LESSON_JS = _html.load_lesson_js(__file__)


def render() -> str:
    controls_html = """
<section class="controls">
  <h2>Semijoin shape controls</h2>
  <div class="control-grid">
    <div class="control">
      <label for="outer_rows">Outer rows (customers): <span class="value-pill" data-pill-for="outer_rows">50000</span></label>
      <input type="range" id="outer_rows" name="outer_rows" min="1000" max="5000000" step="1000" value="50000">
      <div class="hint">Rows in the driving table — the customers the IN subquery is checked against.</div>
    </div>
    <div class="control">
      <label for="inner_matches">Matching inner rows per outer row (orders &gt; $1000): <span class="value-pill" data-pill-for="inner_matches">8</span></label>
      <input type="range" id="inner_matches" name="inner_matches" min="1" max="200" step="1" value="8">
      <div class="hint">Average qualifying orders per customer. A plain join reads them all; FirstMatch reads only until the first one.</div>
    </div>
  </div>
</section>
"""

    query_card_html = _html.query_card(
        sql=(
            "SELECT c.customer_id, c.name\n"
            "FROM   customers c\n"
            "WHERE  c.customer_id IN (\n"
            "  SELECT o.customer_id\n"
            "  FROM   orders o\n"
            "  WHERE  o.total > 1000\n"
            ");"
        ),
        note=(
            "IN (subquery) asks does-it-exist, not how-many. MySQL rewrites "
            "it as a semijoin and runs the FirstMatch strategy: the moment "
            "the first order over $1000 is found for a customer, that "
            "customer is emitted and the inner scan short-circuits — the "
            "remaining orders for that customer are never read."
        ),
    )

    explainer_html = _html.explainer(
        "What you'll see in the animation",
        [
            "Left = the driving customers (outer rows). Orange = the "
            "customer currently being checked. Right = that customer's "
            "orders over $1000 (the inner rows the subquery scans).",
            "Act 1 — naive inner join: for each customer, EVERY matching "
            "order arcs into the probe panel. Alice with 3 orders reads "
            "all 3. The 'rows read' counter climbs fast — most of that "
            "work is wasted for an IN check.",
            "Act 2 — FirstMatch early-out: the same query as a semijoin. "
            "The FIRST matching order arcs in, the customer is emitted "
            "with a green check, and the inner scan STOPS. The skipped "
            "orders are drawn greyed-out — MySQL never reads them.",
            "The two 'rows read' counters diverge — that divergence IS "
            "the lesson. Naive = outer_rows × matches; FirstMatch = "
            "outer_rows × 1 (one probe stops each inner scan).",
            "Watch the phase label: it names the exact customer and order "
            "at every step ('Alice id=1 -> order #104 $1500 matches -> "
            "emit Alice, stop scanning her orders').",
        ],
    )

    stage_html = f"""
<section class="stage">
  {query_card_html}
  {explainer_html}
  {_html.stage_toolbar("Ready — press Play")}
  <div class="stage-with-phases">
    <svg id="fm-svg" viewBox="0 0 800 420" xmlns="http://www.w3.org/2000/svg"></svg>
    {_html.phase_nav()}
  </div>
</section>
"""

    ht = _html.help_tip
    readout_html = f"""
<section class="readout">
  <h2>FirstMatch cost model</h2>
  <div class="readout-grid">
    <div class="item"><p class="label">Outer rows {ht("Rows from the driving (customers) side. The semijoin runs the existence check once per outer row.")}</p><p class="value" id="out-outer">—</p></div>
    <div class="item"><p class="label">Matching inner rows each {ht("Average orders over $1000 per customer. A plain join reads all of them; FirstMatch reads only until the first.")}</p><p class="value" id="out-inner">—</p></div>
    <div class="item"><p class="label">Rows read — naive join {ht("Inner rows a plain inner join would read: outer_rows × matches. All of it is wasted past the first match, because IN only needs existence.")}</p><p class="value hot" id="out-naive">—</p></div>
    <div class="item"><p class="label">Rows read — FirstMatch {ht("Inner rows FirstMatch reads: about outer_rows × 1. The inner scan short-circuits on the first match for each outer row.")}</p><p class="value ok" id="out-firstmatch">—</p></div>
    <div class="item"><p class="label">Inner reads saved {ht("naive − FirstMatch. These are inner-row reads FirstMatch skips because IN asks does-it-exist, not how-many.")}</p><p class="value ok" id="out-saved">—</p></div>
    <div class="item"><p class="label">Complexity {ht("FirstMatch turns the inner cost per outer row from 'read all matches' into 'read until the first match' — roughly constant per outer row.")}</p><p class="value">O(n · 1)</p></div>
  </div>
  <div class="explanation" id="out-explanation"></div>
  <div class="complexity-chart">
    <p class="chart-title">Inner rows read: naive join vs FirstMatch (log-log)</p>
    <svg id="complexity-chart" viewBox="0 0 560 200" xmlns="http://www.w3.org/2000/svg"></svg>
  </div>
</section>
"""

    learn_more_html = """
<details class="learn-more">
  <summary>Learn more — FirstMatch among the four semijoin strategies</summary>
  <div class="body">
    <p>MySQL and MariaDB rewrite <code>IN (SELECT …)</code> and
    <code>EXISTS (SELECT …)</code> subqueries as <strong>semijoins</strong>
    — a join that keeps an outer row if <em>at least one</em> inner row
    matches, without duplicating the outer row for every match. That
    "at least one" is the whole point: the subquery asks
    <em>does a matching row exist?</em>, never <em>how many?</em>.</p>

    <p><strong>FirstMatch</strong> (this lesson) exploits exactly that.
    While scanning the inner side for the current outer row, the instant
    the first qualifying inner row is found, the outer row is emitted and
    the inner scan for that outer row <em>stops</em>. No further inner
    rows are read for that customer. It behaves like a nested loop with
    an early <code>break</code> the moment a match appears — which is why
    it is the cheapest strategy when each outer row has many matches you
    would otherwise read for nothing.</p>

    <p>The other three strategies handle the cases FirstMatch can't:</p>
    <ul>
      <li><strong>LooseScan</strong> — walks the inner index and skips
      duplicate key values, feeding only distinct keys to the join.
      Needs a suitable index on the inner side.</li>
      <li><strong>Materialization</strong> — runs the subquery once into
      a temp table, then probes it per outer row. Good when the subquery
      result is small and reusable.</li>
      <li><strong>DuplicateWeedout</strong> — runs the full join, then
      removes duplicate outer rows via a temp table keyed on the outer
      rowid. The most general fallback — it always works.</li>
    </ul>

    <p>FirstMatch is controlled by
    <code>optimizer_switch='firstmatch=on'</code> (enabled by default in
    MySQL 8.4 and MariaDB 11.4). The optimizer picks it when the outer
    row's existence check can be satisfied by the first inner match and
    reading the rest would be wasted work.</p>

    <p>Practical takeaway: if <code>EXPLAIN</code> shows
    <code>FirstMatch(customers)</code> (or the JSON tree shows a semijoin
    with a first-match strategy), MySQL is <em>already</em> short-circuiting
    the inner scan for you. If instead you see the full inner side being
    read per outer row, an index on the inner join/filter column is what
    lets the optimizer keep the per-outer cost near 1.</p>

    <p>Sources: MySQL 8.4 Reference Manual §10.2.2.1 "Optimizing IN and
    EXISTS Subquery Predicates with Semijoin Transformations"; MariaDB
    Knowledge Base "Semijoin Subquery Optimizations" / "FirstMatch
    Strategy". Behavioural claim (inner scan stops on first match) should
    be verified against the FirstMatch semijoin strategy in
    <code>sql/</code> before shipping any hard cost number.</p>
  </div>
</details>
"""

    return _html.render_page(
        lesson_id="semijoin_firstmatch",
        title="Semijoin FirstMatch — stop looking after the first match",
        subtitle=(
            "IN (subquery) asks does-it-exist, not how-many. Watch MySQL "
            "short-circuit the inner scan the moment one matching order "
            "appears — the naive-join and FirstMatch counters diverge."
        ),
        version_chip="MySQL 8.4 • MariaDB 11.4",
        controls_html=controls_html,
        stage_html=stage_html,
        readout_html=readout_html,
        learn_more_html=learn_more_html,
        lesson_js=_LESSON_JS,
    )

File: myflames/teach/join_family/semijoin_firstmatch.js
function firstMatchCost(outerRows, innerMatches) {
  var outer = Math.max(1, outerRows);
  var matches = Math.max(1, innerMatches);
  // Naive inner join reads every matching inner row for every outer row.
  var naiveInnerReads = outer * matches;
  // FirstMatch stops on the first match: ~1 inner read per outer row.
  var firstMatchInnerReads = outer * 1;
  return {
    outerRows: outer,
    innerMatches: matches,
    naiveInnerReads: naiveInnerReads,
    firstMatchInnerReads: firstMatchInnerReads,
    innerReadsSaved: naiveInnerReads - firstMatchInnerReads
  };
}

var W = 800, H = 420;
var stage = null;

// Sample data: customers (outer) and, for each, their orders over $1000
// (the inner rows the IN subquery scans). Alice/Initech have many; that
// fan-out is where FirstMatch pays off.
var CUSTOMERS = [
  {id: 1, name: "Alice"},
  {id: 2, name: "Bob"},
  {id: 3, name: "Carol"},
  {id: 4, name: "Dave"},
  {id: 5, name: "Eve"}
];

// order_id + total for each customer's qualifying orders (total > 1000).
var ORDERS_BY_CUSTOMER = {
  1: [{oid: 104, total: 1500}, {oid: 118, total: 2200}, {oid: 131, total: 3100}],
  2: [{oid: 102, total: 1200}],
  3: [{oid: 107, total: 1100}, {oid: 109, total: 4800}, {oid: 121, total: 1900}, {oid: 140, total: 2500}],
  4: [],
  5: [{oid: 106, total: 5000}, {oid: 133, total: 1300}]
};

// --- palette (unified with nested_loop / hash lessons) ----------------
var OUTER_BG_REST   = "#ffedd5";
var OUTER_BG_DRIVE  = "#fde68a";
var OUTER_BG_DONE   = "#dcfce7";   // emitted (matched)
var OUTER_BG_NONE   = "#f3f4f6";   // checked, no match
var OUTER_STROKE    = "#fdba74";
var OUTER_STROKE_DR = "#f59e0b";
var OUTER_STROKE_OK = "#16a34a";
var PROBE_BG        = "#eff6ff";
var PROBE_STROKE    = "#7dd3fc";
var PILL_FILL       = "#bae6fd";
var PILL_STROKE     = "#0284c7";
var PILL_MATCH_FILL = "#bbf7d0";
var PILL_MATCH_STK  = "#16a34a";
var PILL_SKIP_FILL  = "#e5e7eb";   // greyed: never read
var PILL_SKIP_STK   = "#9ca3af";
var MATCH_TEXT      = "#0c4a6e";
var SKIP_TEXT       = "#9ca3af";

function buildStage() {
  var svg = document.getElementById("fm-svg");
  while (svg.firstChild) svg.removeChild(svg.firstChild);

  var leftX = 24, topY = 52, colW = 300;
  var rightX = 470, rightW = 300;
  var rowH = 46;

  var outerLbl = anim.svgEl("text", {
    x: leftX, y: 26, "font-size": 13, "font-weight": 700, fill: "#7c2d12"
  });
  outerLbl.textContent = "Outer (driving): customers checked by IN (…)";
  svg.appendChild(outerLbl);

  var innerLbl = anim.svgEl("text", {
    x: rightX, y: 26, "font-size": 13, "font-weight": 700, fill: "#0c4a6e"
  });
  innerLbl.textContent = "Inner scan: this customer's orders > $1000";
  svg.appendChild(innerLbl);

  // Outer-row boxes.
  var outerRows = [];
  for (var i = 0; i < CUSTOMERS.length; i++) {
    var y = topY + i * rowH;
    var bg = anim.svgEl("rect", {
      x: leftX, y: y, width: colW, height: rowH - 8, rx: 7, ry: 7,
      fill: OUTER_BG_REST, stroke: OUTER_STROKE, "stroke-width": 1.5
    });
    svg.appendChild(bg);
    var txt = anim.svgEl("text", {
      x: leftX + 12, y: y + 23, "font-size": 11, "font-weight": 600,
      fill: "#7c2d12"
    });
    txt.textContent = "customer_id=" + CUSTOMERS[i].id + " " + CUSTOMERS[i].name;
    svg.appendChild(txt);
    // Per-row verdict badge (right edge of the outer box).
    var badge = anim.svgEl("text", {
      x: leftX + colW - 10, y: y + 23, "text-anchor": "end",
      "font-size": 10.5, "font-weight": 700, fill: "#065f46", opacity: 0
    });
    svg.appendChild(badge);
    outerRows.push({
      bg: bg, txt: txt, badge: badge, data: CUSTOMERS[i],
      y: y, cx: leftX + colW, cy: y + (rowH - 8) / 2
    });
  }

  // Inner probe panel.
  var innerPanel = anim.svgEl("rect", {
    x: rightX, y: topY, width: rightW, height: 210, rx: 8, ry: 8,
    fill: PROBE_BG, stroke: PROBE_STROKE, "stroke-width": 1.5
  });
  svg.appendChild(innerPanel);

  var probeTitle = anim.svgEl("text", {
    x: rightX + 12, y: topY + 22, "font-size": 11.5, "font-weight": 700,
    fill: "#0c4a6e"
  });
  probeTitle.textContent = "Inner rows read for the current customer";
  svg.appendChild(probeTitle);

  var probeLanding = anim.svgEl("g", {id: "probe-landing"});
  svg.appendChild(probeLanding);

  var verdict = anim.svgEl("text", {
    x: rightX + 12, y: topY + 192, "font-size": 12, "font-weight": 700,
    fill: MATCH_TEXT
  });
  verdict.textContent = "";
  svg.appendChild(verdict);

  // Two diverging counters (the story).
  var counterY = topY + 232;
  var naiveCounter = anim.svgEl("text", {
    x: rightX, y: counterY, "font-size": 12, "font-weight": 700, fill: "#b91c1c"
  });
  naiveCounter.textContent = "Naive join inner rows read: 0";
  svg.appendChild(naiveCounter);

  var fmCounter = anim.svgEl("text", {
    x: rightX, y: counterY + 20, "font-size": 12, "font-weight": 700, fill: "#15803d"
  });
  fmCounter.textContent = "FirstMatch inner rows read: 0";
  svg.appendChild(fmCounter);

  var statusLbl = anim.svgEl("text", {
    x: W / 2, y: H - 14, "text-anchor": "middle",
    "font-size": 12.5, "font-weight": 600, fill: "#111827"
  });
  statusLbl.textContent = "";
  svg.appendChild(statusLbl);

  stage = {
    svg: svg,
    outerRows: outerRows,
    probeLanding: probeLanding,
    probePanelX: rightX + 16,
    probePanelY: topY + 44,
    verdict: verdict,
    naiveCounter: naiveCounter,
    fmCounter: fmCounter,
    statusLbl: statusLbl
  };
}

function resetStage() {
  if (!stage) return;
  for (var i = 0; i < stage.outerRows.length; i++) {
    var r = stage.outerRows[i];
    r.bg.setAttribute("fill", OUTER_BG_REST);
    r.bg.setAttribute("stroke", OUTER_STROKE);
    r.badge.setAttribute("opacity", 0);
    r.badge.textContent = "";
  }
  while (stage.probeLanding.firstChild) {
    stage.probeLanding.removeChild(stage.probeLanding.firstChild);
  }
  stage.verdict.textContent = "";
  stage.naiveCounter.textContent = "Naive join inner rows read: 0";
  stage.fmCounter.textContent = "FirstMatch inner rows read: 0";
  stage.statusLbl.textContent = "";
}

// Tweened "this is the current driver" transition + arrival pulse.
function driveRow(tl, row) {
  tl.add({
    from: 0, to: 1, duration: 240, ease: anim.easeOutCubic,
    onUpdate: function(t) {
      row.bg.setAttribute("fill", anim.lerpColor(OUTER_BG_REST, OUTER_BG_DRIVE, t));
      row.bg.setAttribute("stroke", anim.lerpColor(OUTER_STROKE, OUTER_STROKE_DR, t));
    },
    onComplete: function() { anim.arrival(row.bg); }
  });
}

function settleRow(tl, row, matched) {
  var toFill = matched ? OUTER_BG_DONE : OUTER_BG_NONE;
  var toStroke = matched ? OUTER_STROKE_OK : OUTER_STROKE;
  tl.add({
    from: 0, to: 1, duration: 220, ease: anim.easeInCubic,
    onUpdate: function(t) {
      row.bg.setAttribute("fill", anim.lerpColor(OUTER_BG_DRIVE, toFill, t));
      row.bg.setAttribute("stroke", anim.lerpColor(OUTER_STROKE_DR, toStroke, t));
    }
  });
}

var PROBE_SLOT_H = 30;
var PROBE_MAX_SLOTS = 4;

// Spawn one inner-row pill and arc it into the probe panel.
//   read  = true  -> the pill is actually read (blue, or green if it is
//                    the matching row that triggers the early-out).
//   read  = false -> FirstMatch skips it: greyed-out, arrives faded, no
//                    arrival pulse. Communicates "never read".
function spawnInnerPill(tl, row, order, slotIdx, opts) {
  opts = opts || {};
  var read = opts.read !== false;
  var isMatch = !!opts.isMatch;
  var built = false;
  var pill = null;
  var rect = null;
  var label = null;

  var x0 = row.cx + 4;
  var y0 = row.cy;
  var slotClamped = Math.min(slotIdx, PROBE_MAX_SLOTS - 1);
  var x1 = stage.probePanelX + 6;
  var y1 = stage.probePanelY + slotClamped * PROBE_SLOT_H;
  var cx = (x0 + x1) / 2;
  var cy = Math.min(y0, y1) - 34 - slotIdx * 4;
  var pathFn = anim.path(x0, y0, cx, cy, x1, y1);

  var fill = read ? (isMatch ? PILL_MATCH_FILL : PILL_FILL) : PILL_SKIP_FILL;
  var stroke = read ? (isMatch ? PILL_MATCH_STK : PILL_STROKE) : PILL_SKIP_STK;
  var txtColor = read ? MATCH_TEXT : SKIP_TEXT;
  var endOpacity = read ? 1 : 0.55;

  tl.add({
    from: 0, to: 1, duration: read ? 380 : 300, ease: anim.easeInOutQuad,
    onUpdate: function(t) {
      if (!built) {
        built = true;
        pill = anim.svgEl("g", {opacity: read ? 1 : 0.55});
        rect = anim.svgEl("rect", {
          x: -46, y: -12, width: 172, height: 24, rx: 4, ry: 4,
          fill: fill, stroke: stroke, "stroke-width": 1.2,
          "stroke-dasharray": read ? "0" : "4 3"
        });
        label = anim.svgEl("text", {
          x: 40, y: 4, "text-anchor": "middle",
          "font-size": 10, "font-weight": 600, fill: txtColor
        });
        var tag = isMatch ? " ✓ match" : (read ? "" : " — skipped");
        label.textContent =
          "order #" + order.oid + " $" + order.total + tag;
        pill.appendChild(rect);
        pill.appendChild(label);
        stage.probeLanding.appendChild(pill);
      }
      var pt = pathFn(t);
      pill.setAttribute("transform",
        "translate(" + pt.x.toFixed(1) + "," + pt.y.toFixed(1) + ")");
      if (!read) pill.setAttribute("opacity", (0.55 * t).toFixed(2));
    },
    onComplete: function() {
      if (pill) pill.setAttribute("opacity", endOpacity);
      if (rect && read) {
        anim.arrival(rect, {peakWidth: isMatch ? 2.8 : 2.0, durationMs: 260});
      }
    }
  });
}

function clearProbe(tl, headline) {
  tl.call(function() {
    while (stage.probeLanding.firstChild) {
      stage.probeLanding.removeChild(stage.probeLanding.firstChild);
    }
    stage.verdict.textContent = "";
    if (headline) document.getElementById("phase-label").textContent = headline;
  });
}

function buildTimeline() {
  resetStage();
  var tl = anim.timeline();
  var phase = document.getElementById("phase-label");
  var naiveRead = 0;
  var fmRead = 0;

  function setCounters() {
    stage.naiveCounter.textContent = "Naive join inner rows read: " + naiveRead;
    stage.fmCounter.textContent = "FirstMatch inner rows read: " + fmRead;
  }

  // ================= ACT 1 — naive inner join =================
  tl.mark("Act 1 — naive join (reads every match)");
  tl.call(function() {
    phase.textContent =
      "Act 1/2 — a plain inner join reads EVERY matching order per customer (wasted work for IN)";
  });
  tl.delay(500);

  for (var i = 0; i < stage.outerRows.length; i++) {
    (function(idx) {
      var row = stage.outerRows[idx];
      var c = row.data;
      var orders = ORDERS_BY_CUSTOMER[c.id] || [];

      clearProbe(tl,
        "Act 1 — checking " + c.name + " (id=" + c.id + "): reading ALL matching orders");
      driveRow(tl, row);

      for (var j = 0; j < orders.length; j++) {
        (function(jj) {
          spawnInnerPill(tl, row, orders[jj], jj, {read: true, isMatch: false});
          tl.call(function() {
            naiveRead += 1;
            setCounters();
            phase.textContent = "Act 1 — " + c.name + " (id=" + c.id +
              ") reads order #" + orders[jj].oid + " $" + orders[jj].total +
              " (" + (jj + 1) + "/" + orders.length + ")";
          });
          tl.delay(90);
        })(j);
      }

      tl.call(function() {
        var n = orders.length;
        if (n === 0) {
          stage.verdict.textContent = c.name + " (id=" + c.id + ") → no orders > $1000";
          row.badge.textContent = "no match";
          row.badge.setAttribute("fill", "#6b7280");
        } else {
          stage.verdict.textContent = c.name + " (id=" + c.id + ") → read all " +
            n + (n === 1 ? " order" : " orders");
          row.badge.textContent = "read " + n;
          row.badge.setAttribute("fill", "#b91c1c");
        }
        row.badge.setAttribute("opacity", 1);
      });
      settleRow(tl, row, orders.length > 0);
      tl.delay(650);
    })(i);
  }

  tl.call(function() {
    phase.textContent =
      "Act 1 done — naive join read " + naiveRead +
      " inner rows. But IN only needs to know a match EXISTS. Watch FirstMatch…";
    stage.statusLbl.textContent =
      "Every order past the first was wasted work for the IN check.";
  });
  tl.delay(1100);

  // ================= ACT 2 — FirstMatch early-out =================
  tl.mark("Act 2 — FirstMatch (stops on first)");
  tl.call(function() {
    // Reset outer rows to rest so the second pass reads clean.
    for (var k = 0; k < stage.outerRows.length; k++) {
      var rr = stage.outerRows[k];
      rr.bg.setAttribute("fill", OUTER_BG_REST);
      rr.bg.setAttribute("stroke", OUTER_STROKE);
      rr.badge.setAttribute("opacity", 0);
    }
    phase.textContent =
      "Act 2/2 — same query as a semijoin: stop the inner scan on the FIRST match";
  });
  tl.delay(500);

  for (var m = 0; m < stage.outerRows.length; m++) {
    (function(idx) {
      var row = stage.outerRows[idx];
      var c = row.data;
      var orders = ORDERS_BY_CUSTOMER[c.id] || [];

      clearProbe(tl,
        "Act 2 — checking " + c.name + " (id=" + c.id + "): scan until the first match");
      driveRow(tl, row);

      if (orders.length === 0) {
        // No match: the inner side is scanned but finds nothing.
        tl.call(function() {
          fmRead += 1;               // one probe to discover "no rows"
          setCounters();
          stage.verdict.textContent = c.name + " (id=" + c.id + ") → no match — not in result";
          row.badge.textContent = "no match";
          row.badge.setAttribute("fill", "#6b7280");
          row.badge.setAttribute("opacity", 1);
          phase.textContent = "Act 2 — " + c.name + " (id=" + c.id +
            ") has no order > $1000 → excluded from the IN result";
        });
        settleRow(tl, row, false);
        tl.delay(650);
        return;
      }

      // First order: read it, it matches -> emit + STOP.
      spawnInnerPill(tl, row, orders[0], 0, {read: true, isMatch: true});
      tl.call(function() {
        fmRead += 1;
        setCounters();
        phase.textContent = "Act 2 — " + c.name + " (id=" + c.id +
          ") → order #" + orders[0].oid + " $" + orders[0].total +
          " matches → emit " + c.name + ", STOP scanning";
        stage.verdict.textContent = c.name + " (id=" + c.id +
          ") → first match on order #" + orders[0].oid + " — emit, short-circuit";
        row.badge.textContent = "✓ emit (read 1)";
        row.badge.setAttribute("fill", "#15803d");
        row.badge.setAttribute("opacity", 1);
      });
      tl.delay(140);

      // Remaining orders: greyed, never read.
      for (var j = 1; j < orders.length; j++) {
        (function(jj) {
          spawnInnerPill(tl, row, orders[jj], jj, {read: false, isMatch: false});
          tl.delay(55);
        })(j);
      }
      tl.call(function() {
        if (orders.length > 1) {
          phase.textContent = "Act 2 — " + c.name + "'s other " +
            (orders.length - 1) + " matching order" +
            (orders.length - 1 === 1 ? "" : "s") +
            " are SKIPPED (greyed) — MySQL never reads them";
        }
      });
      settleRow(tl, row, true);
      tl.delay(650);
    })(m);
  }

  tl.mark("Consequence");
  tl.call(function() {
    phase.textContent =
      "FirstMatch read " + fmRead + " inner rows vs " + naiveRead +
      " for the naive join — IN asks does-it-exist, so the inner scan short-circuits";
    stage.statusLbl.textContent =
      "Saved " + (naiveRead - fmRead) + " inner-row reads on 5 customers. " +
      "At scale the gap is outer_rows × (matches − 1).";
  });
  tl.delay(700);
  return tl;
}

function buildCurrentTimeline() {
  return buildTimeline();
}

function resetAnim() {
  resetStage();
  document.getElementById("phase-label").textContent = "Ready — press Play";
}

function renderChart(innerMatches, currentOuterRows) {
  anim.complexityChart({
    svgId: "complexity-chart",
    width: 560, height: 200,
    xMin: 1000, xMax: 5e6,
    xLabel: "Outer (customer) rows", yLabel: "Inner rows read",
    curves: [
      { label: "Naive join: O(n · matches)", color: "#b91c1c",
        fn: function(n) { return n * innerMatches; } },
      { label: "FirstMatch: O(n · 1)", color: "#15803d",
        fn: function(n) { return n; } }
    ],
    current: { x: currentOuterRows },
    xSlider: "outer_rows",
    xSliderTransform: function(xVal) {
      return Math.max(1000, Math.round(xVal / 1000) * 1000);
    }
  });
}

function recompute() {
  var c = teachRuntime.readControls();
  var cost = firstMatchCost(c.outer_rows, c.inner_matches);
  document.getElementById("out-outer").textContent =
    teachRuntime.formatInt(cost.outerRows);
  document.getElementById("out-inner").textContent =
    teachRuntime.formatInt(cost.innerMatches);
  document.getElementById("out-naive").textContent =
    teachRuntime.formatInt(cost.naiveInnerReads);
  document.getElementById("out-firstmatch").textContent =
    teachRuntime.formatInt(cost.firstMatchInnerReads);
  document.getElementById("out-saved").textContent =
    teachRuntime.formatInt(cost.innerReadsSaved);
  document.getElementById("out-explanation").textContent =
    "A plain inner join would read every matching order for each customer: " +
    teachRuntime.formatInt(cost.outerRows) + " customers × " +
    teachRuntime.formatInt(cost.innerMatches) + " matches = " +
    teachRuntime.formatInt(cost.naiveInnerReads) + " inner rows read. " +
    "But IN (subquery) only asks whether a match EXISTS, not how many. " +
    "FirstMatch stops the inner scan on the first match for each customer, " +
    "so it reads about " + teachRuntime.formatInt(cost.firstMatchInnerReads) +
    " inner rows — saving " +
    teachRuntime.formatInt(cost.innerReadsSaved) +
    " reads. The saving is outer_rows × (matches − 1): the more " +
    "orders each customer has, the more FirstMatch skips.";
  buildStage();
  resetStage();
  renderChart(cost.innerMatches, cost.outerRows);
}

buildStage();
teachRuntime.wire(recompute);
teachRuntime.wireToolbar({
  build: buildCurrentTimeline,
  reset: resetAnim
});
teachRuntime.wirePhaseNav("phase-nav", {
  build: buildCurrentTimeline,
  reset: resetAnim
});
